### ISSUE TITLE
Compiling of custom commands gets stuck

### ISSUE THREAD

**Issue by dirkziegenbalg** [texstudio-org/texstudio#690]
### Environment
- TeXstudio: 2.12.16 (git 2.12.16-52-ga7fd0a3) starting with 2.12.16
- Qt: 5.12.1
- OS: Windows 10
- TeX distribution: miktex

### Expected behavior

custom command should close an eventually open PDF in Adobe Acrobat and start compiling afterwards

### Actual behavior

Command gets stuck after sending the DDE command to Acrobat and the compiling interface (buttons, etc.) is deactivated. The compilation command is not executed. Calling the custom commands separately doesn't cause any issues.  It's not possible to start any compilation until Texstudio is restarted.

The command runs well with Texstudio 2.12.14. Issues started with version 2.12.16 and is still present with the latest git version.

### How to reproduce
Use the following custom build command:

close-adobe:Close ----> dde:///"C:\Program Files (x86)\Adobe\Acrobat DC\Acrobat\Acrobat.exe":acroviewA19/control/[DocClose("?am.pdf")][FileClose("?am.pdf")]"

create:create_with_close ---->   txs:///close-adobe | txs:///pdftex_user

pdftex_user:create_pdf ----> pdflatex.exe -shell-escape -synctex=1 -interaction=nonstopmode %.tex

**Comment by MeanSquaredError** [comment]
Did you build it on Windows yourself or did you use a custom build?
I would say that the dde commands are the prime suspects, although I do not have a TexStudio Windows build for testing.

You can try replacing pdflatex.exe with a no-op (empty) batch file. E.g. something like this 
pdftex_user:create_pdf ----> noop.bat, where noop.bat is an empty file. And then see if the UI lockup still persists. If the lockup persists then DDE is the suspect.

**Comment by dirkziegenbalg** [comment]
I downloaded the file from bintray (build from 11.8.19) but the issue 
also exists in the official 2.12.16 version.

Just checked again with the build from 11.8.19 and the issue persists 
with the noop.bat. TXS UI is till locked up and the batch file is not 
executed.

When I reinstall version 2.12.14 every runs fine as expected. The batch 
file as well as pdflatex are executed.


Am 16.08.2019 um 12:43 schrieb MeanSquaredError:

**Comment by MeanSquaredError** [comment]
Hi,
Does the UI lockup happen with TexStudio 2.12.16 if you replace the command in close-adobe:Close with an empty noop.bat? That is 

close-adobe:Close ----> noop.bat

**Comment by dirkziegenbalg** [comment]
Hi,

no. In this case TXS tries to execute the batch file but fails and the 
total command is stopped. If I add a print command, the batch file runs 
and the follow-up command is executed as well.

I just recognized that pdftex is executed when I close TXS when I use 
the DDE command and the UI is locked up. Furthermore, there seems to be 
a little difference between the official 2.12.16 and the version from 
bintray (11.08.2019). For the official one it seems that TXS gets stuck 
but didn't lock up the UI. I also can execute other commands. For the 
bintray version I cannot execute any other commands and the buttons for 
e.g. the green arrow for compiling is deactivated.

To me it seems as TXS doesn't correctly recognizes that the DDE command 
execution is completed.

Am 16.08.2019 um 16:15 schrieb MeanSquaredError:

**Comment by MeanSquaredError** [comment]
I think I figured it out. There is a race condition which prevents the proper unlocking of buttons. It goes like this :

BuildManager::runCommandInternal calls p->startCommand which in turn calls manager->executeDDE()
The executeDDE function finishes quickly and calls waitForProcess immediately.

After that BuildManager::runCommandInternal calls waitForProcess(p) which does the following

connect(p, SIGNAL(processFinished()), &loop, SLOT(quit())); // This starts listening for processFinished event
emit buildRunning(true); // This disables the UI buttons
loop.exec(); // Loops until processFinished is signaled
emit buildRunning(false); // This enables the UI buttons

So basically the processFinished  is signaled before we start listening for it and  we never get the signal and loop.exec() never terminates.

I don't have the time to fix it properly at the moment (mostly because I don't have access to VS 2017 and latest QT does not work well with VS 2015), but I guess some developer who can build it on Windows can do the proper fix. If nobody does it, I will be able to work on it in about 3 weeks.

**Comment by dirkziegenbalg** [comment]
that sounds reasonable and actually that is the kind of behavior I 
expected somehow.

It's not that urgent for me. Currently I'm running my production system 
with 2.12.14 where everything runs as expected.

Thanks for the fast troubleshooting!

Am 19.08.2019 um 00:52 schrieb MeanSquaredError:

**Comment by MeanSquaredError** [comment]
Actually I see that above I messed up the sequence slightly

manager->executeDDE() actually calls processFinished (and not waitForProcess). It is BuildManager::runCommandInternal that calls waitForProcess. It does not call waitForProcess when the executed command is the last in the list. So when the dde:/// command is the only command it is not waited for and everything works correctly.

I might be able to provide a patch today if I am able to get Qt working with MinGW today.

**Comment by dirkziegenbalg** [comment]
Just tested the bintray version from 20.08. and everything runs smooth now! Thanks for the fast bug fixing!
